File: modules/arithmetic/doc/average.py
[ ## this file was manually modified by jt
    {
     'functor' : {
         'arity' : '2',
         'call_types' : [],
         'ret_arity' : '0',
         'rturn' : {
             'default' : 'T',
            },
         'simd_types' : ['real_'],
         'type_defs' : [],
         'types' : ['real_', 'signed_int_', 'unsigned_int_'],
        },
     'info' : 'manually modified',
     'unit' : {
         'global_header' : {
             'first_stamp' : 'modified by jt the 28/11/2010',
             'included' : ['#include<nt2/include/functions/round.hpp>'],
             'cover_included' : ['#include<nt2/include/functions/iround.hpp>'],
             'notes' : ['for integer values average does not coincide with (a0+a1)/2 by at most one unit.'],
             'stamp' : 'modified by jt the 13/12/2010',
            },
         'ranges' : {
             'real_' : [['T(-100)', 'T(100)'], ['T(-100)', 'T(100)']],
             'signed_int_' : [['T(-100)', 'T(100)'], ['T(-100)', 'T(100)']],
             'unsigned_int_' : [['T(0)', 'T(100)'], ['T(0)', 'T(100)']],
            },
         'specific_values' : {
             'default' : {
                },
             'real_' : {
                 'nt2::Inf<T>()' : 'nt2::Inf<T>()',
                 'nt2::Minf<T>()' : 'nt2::Minf<T>()',
                 'nt2::Mone<T>()' : 'nt2::Mone<T>()',
                 'nt2::Nan<T>()' : 'nt2::Nan<T>()',
                 'nt2::One<T>()' : 'nt2::One<T>()',
                 'nt2::Zero<T>()' : 'nt2::Zero<T>()',
                 'nt2::Zero<T>(),nt2::One<T>()' : 'nt2::Half<T>()',   
                },
             'signed_int_' : {
                 'nt2::Mone<T>()' : 'nt2::Mone<T>()',
                 'nt2::One<T>()' : 'nt2::One<T>()',
                 'nt2::Zero<T>()' : 'nt2::Zero<T>()',
                 'nt2::Zero<T>(),nt2::One<T>()' : 'nt2::Zero<T>()',   
                },
             'unsigned_int_' : {
                 'nt2::One<T>()' : 'nt2::One<T>()',
                 'nt2::Zero<T>()' : 'nt2::Zero<T>()',
                 'nt2::Zero<T>(),nt2::One<T>()' : 'nt2::Zero<T>()',   
                },
            },
         'verif_test' : {
             'property_call' : {
                 'default' : ['nt2::average(a0,a1)'],
                },
             'property_value' : {
                 'default' : ['nt2::iround((a0+a1)/2)'],
                 'real_'   : ['(a0+a1)/2'],   
                },
             'ulp_thresh' : {
                 'default' : ['1'],
                 'real_' : ['0'],
                },
            },
        },
     'version' : '0.1',
    },
]

File: modules/boost/simd/base/unit/arithmetic/scalar/average.cpp
//==============================================================================
//         Copyright 2003 - 2012   LASMEA UMR 6602 CNRS/Univ. Clermont II
//         Copyright 2009 - 2012   LRI    UMR 8623 CNRS/Univ Paris Sud XI
//
//          Distributed under the Boost Software License, Version 1.0.
//                 See accompanying file LICENSE.txt or copy at
//                     http://www.boost.org/LICENSE_1_0.txt
//==============================================================================
/// for integer values average does not,coincide with (a0+a1)/2 by at most one unit.
#include <boost/simd/arithmetic/include/functions/average.hpp>
#include <boost/dispatch/functor/meta/call.hpp>
#include <nt2/sdk/unit/tests/relation.hpp>
#include <nt2/sdk/unit/tests/type_expr.hpp>
#include <nt2/sdk/unit/module.hpp>
#include <boost/simd/include/constants/two.hpp>
#include <boost/simd/include/constants/one.hpp>
#include <boost/simd/include/constants/mone.hpp>
#include <boost/simd/include/constants/inf.hpp>
#include <boost/simd/include/constants/minf.hpp>
#include <boost/simd/include/constants/nan.hpp>
#include <boost/simd/sdk/config.hpp>


NT2_TEST_CASE_TPL ( average_real,  BOOST_SIMD_REAL_TYPES)
{

  using boost::simd::average;
  using boost::simd::tag::average_;
  typedef typename boost::dispatch::meta::call<average_(T,T)>::type r_t;
  typedef T wished_r_t;

  // return type conformity test
  NT2_TEST_TYPE_IS(r_t, wished_r_t);

  // specific values tests
#ifndef BOOST_SIMD_NO_INVALIDS
  NT2_TEST_EQUAL(average(boost::simd::Inf<T>(), boost::simd::Inf<T>()), boost::simd::Inf<T>());
  NT2_TEST_EQUAL(average(boost::simd::Minf<T>(), boost::simd::Minf<T>()), boost::simd::Minf<T>());
  NT2_TEST_EQUAL(average(boost::simd::Nan<T>(), boost::simd::Nan<T>()), boost::simd::Nan<T>());
#endif
  NT2_TEST_EQUAL(average(boost::simd::Mone<T>(), boost::simd::Mone<T>()), boost::simd::Mone<T>());
  NT2_TEST_EQUAL(average(boost::simd::One<T>(), boost::simd::One<T>()), boost::simd::One<T>());
  NT2_TEST_EQUAL(average(boost::simd::Zero<T>(), boost::simd::Zero<T>()), boost::simd::Zero<T>());
} // end of test for floating_

NT2_TEST_CASE_TPL ( average_signed_int,  BOOST_SIMD_INTEGRAL_SIGNED_TYPES)
{

  using boost::simd::average;
  using boost::simd::tag::average_;
  typedef typename boost::dispatch::meta::call<average_(T,T)>::type r_t;
  typedef T wished_r_t;


  // return type conformity test
  NT2_TEST_TYPE_IS(r_t, wished_r_t);

  // specific values tests
  NT2_TEST_EQUAL(average(boost::simd::Mone<T>(), boost::simd::Mone<T>()), boost::simd::Mone<T>());
  NT2_TEST_EQUAL(average(boost::simd::One<T>(), boost::simd::One<T>()), boost::simd::One<T>());
  NT2_TEST_EQUAL(average(boost::simd::Zero<T>(), boost::simd::Zero<T>()), boost::simd::Zero<T>());
  NT2_TEST_EQUAL(average(boost::simd::One<T>(), boost::simd::Zero<T>()), boost::simd::Zero<T>());
  NT2_TEST_EQUAL(average(boost::simd::One<T>(), boost::simd::Two<T>()), boost::simd::One<T>());
} // end of test for signed_int_

NT2_TEST_CASE_TPL ( average_unsigned_int,  BOOST_SIMD_UNSIGNED_TYPES)
{

  using boost::simd::average;
  using boost::simd::tag::average_;
  typedef typename boost::dispatch::meta::call<average_(T,T)>::type r_t;
  typedef T wished_r_t;


  // return type conformity test
  NT2_TEST_TYPE_IS(r_t, wished_r_t);

  // specific values tests
  NT2_TEST_EQUAL(average(boost::simd::One<T>(), boost::simd::One<T>()), boost::simd::One<T>());
  NT2_TEST_EQUAL(average(boost::simd::Zero<T>(), boost::simd::Zero<T>()), boost::simd::Zero<T>());
} // end of test for unsigned_int_
